app/lib/shared/deps_graph.dart

```diff
@@ -0,0 +1,218 @@
+    if (_deps.putIfAbsent(node, _newSet).add(dep)) {
+      _fixpointNode(node);
+    }
+  }
+
+  Set<String> transitiveNodes(String node) => _deps.putIfAbsent(node, _newSet);
+
+  void _fixpointNode(String node) {
+    final Set<String> directs = _deps[node];
+    final temp = <Set<String>>[];
+    for (;;) {
+      final int before = directs.length;
+      temp.clear();
+      for (final String direct in directs) {
+        if (node != direct) {
+          final Set<String> transitives = _deps[direct];
+          if (transitives?.isNotEmpty == true) {
+            temp.add(transitives);
+          }
+        }
+      }
+      for (final Set<String> transitives in temp) {
+        directs.addAll(transitives);
+      }
+      if (before == directs.length) break;
+    }
+  }
+
+  static Set<String> _newSet() => new Set<String>();
+}
+
+class PackageDependencyBuilder {
+  static const Duration pollingInterval = const Duration(seconds: 10);
+  static const String devPrefix = 'dev/';
+
+  final DatastoreDB db;
+
+  // The forward (transitive) dependencies could be easily calculated, but
+  // right now we have no need for them.
+  //final TransitiveDependencyGraph normalDeps = new TransitiveDependencyGraph();
+
+  final TransitiveDependencyGraph reverseDeps = new TransitiveDependencyGraph();
+  final String Function(String) canonicalize = new StringCanonicalization().canonicalize;
+
+  DateTime _lastTs;
+
+  PackageDependencyBuilder(this.db);
+
+  Future scanExistingPackageGraph() async {
+    final sw = new Stopwatch()..start();
+    for (;;) {
+      _logger.info('Scanning existing package graph');
+      try {
+        // We scan from oldest to newest and therefore keep [_lastTs] always
+        // increasing.
+        final query = dbService.query(PackageVersion)..order('created');
+        await for (PackageVersion pv in query.run()) {
+          _addPackageVersion(pv);
+          _lastTs = pv.created;
+        }
+      } catch (e, s) {
+        _logger.severe(e, s);
+        continue;
+      }
+      _logger.info('Scanned package graph in ${sw.elapsed}');
+      return;
+    }
+  }
+
+  Stream<Set<String>> run() async* {
+    _logger.info('Monitoring new package uploads & trigger analysis');
+    for (;;) {
+      try {
+        final query = dbService.query(PackageVersion)
+          ..filter('created >', _lastTs)
+          ..order('created');
+        await for (PackageVersion pv in query.run()) {
+          // Emit a set of packages depending on this one.
+          _addPackageVersion(pv);
+          yield _affectedPackages(pv.package);
+
+          _lastTs = pv.created;
+        }
+      } catch (e, s) {
+        _logger.severe(e, s);
+      }
+      await new Future.delayed(pollingInterval);
+    }
+  }
+
+  void _addPackageVersion(PackageVersion pv) {
+    final String package = pv.package;
+    final Map pubspec = pv.pubspec.asJson;
+
+    final Set<String> depsSet = new Set<String>();
+    final normalDeps = pubspec['dependencies'];
```

Comment:
I'd put these in the `Pubspec` class (which hopefully will get merged with `pana`'s `Pubspec` class), let's remove the duplicate functions here and there...